source/_posts/2018-05-31-automating-your-infrastructure-deployments.markdown

```diff
@@ -0,0 +1,88 @@
+
+Now that we have seen the pros and cons of automating your infrastructure let's see a concrete example of a possible way of achieving that.
+
+# CI/CD with Terraform and CircleCI
+
+I created an [example repository](https://github.com/fedekau/terraform-with-circleci-example) that you can use to kickstart your project or to make your current project better with some of the ideas included there. The repository includes a detailed readme of why I took all the decisions included there.
+
+The main purpose is to show one of the many possible ways you could manage your infrastructure using an Infrastructure as Code (IAC) tool like [Terraform](https://www.terraform.io/) and a continuous integration tool like [CircleCI](https://circleci.com).
+
+I have followed many of the good practices described in the book [Terraform: Up & Running](https://www.terraformupandrunning.com/) and in the [CircleCI 2.0 Documentation](https://circleci.com/docs/2.0/). I will also assume you have some knowledge about those tools and [Amazon Web Services](https://aws.amazon.com) in general, no need to be an expert.
+
+## Suggested workflow
+
+Assuming you have two environments, `production` and `staging`, when a new feature is requested you branch from `staging`, commit the code and open a PR to the `staging` branch, when you do, CircleCI will run two jobs, one for linting and one that will plan the changes to your `staging` infrastructure so you can review them (see image below).
+
+![Image of PR creation jobs](https://raw.githubusercontent.com/fedekau/terraform-with-circleci-example/staging/.images/pr.png)
+
+Once you merge the PR, if everything goes as planned CircleCI will run your jobs and will automatically deploy your infrastructure!
+
+![Image of jobs after staging merge](https://raw.githubusercontent.com/fedekau/terraform-with-circleci-example/staging/.images/staging-merge.png)
+
+After you have tested your infrastructure in `staging` you just need to open a PR from `staging` into `master` to "promote" you infrastructure into `production`.
+
+In this case we want someone to manually approve the release into master, so after you merge you need to tell CircleCI that it can proceed and it will deploy the infrastructure after the confirmation.
+
+![Image of jobs after master merge](https://raw.githubusercontent.com/fedekau/terraform-with-circleci-example/staging/.images/master-merge.png)
+
+## Improvements
+
+This is a very basic workflow and there are many things that must (or can) be improved. Here are some of them:
+
+- The most important improvement that needs to be done in my opinion is to add a way to test the infrastructure for each PR before merging them into `staging`. Something like the [Heroku Review Apps](https://devcenter.heroku.com/articles/github-integration-review-apps). If you use Terraform modules and you write them in a generic way you could very easily implement this.
```

Comment:
I would add commas before and after _in my opinion_: _The most important improvement that needs to be done, in my opinion, is to add_